sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -964,6 +965,155 @@ object CombineFilters extends Rule[LogicalPlan] with PredicateHelper {
+ * to the partial averages of the inner aggregate
+ *
+ * Min an Max can be folded in the case described above, or if they are referencing
+ * the group by columns, as they can safely be computed just using the set of
+ * unique values.
+ */
+object CombineAggregates extends Rule[LogicalPlan] with PredicateHelper {
+
+  /**
+   * The aggregate expression list includes both aggregate expressions and
+   * the projected group by keys, this filters out the aggregate expressions
+   * in the list leaving just the group by keys. It also unwraps aliases to
+   * just give a list of the projected grouping expressions themselves.
+   */
+  def justProjectedGroupExprs(aggExprs: Seq[NamedExpression],
+                   groupExprs: Seq[Expression]): Seq[NamedExpression] = {
+    aggExprs.filter(namedEx =>
+      groupExprs.exists(_.semanticEquals(unwrapAlias(namedEx)))
+    )
+  }
+
+  def unwrapAlias(ex: Expression): Expression = {
+    if (ex.isInstanceOf[Alias]) ex.children.head
+    else ex
+  }
+
+  /**
+   * Pulls up references to aliases from an earlier operator and replaces them with the
+   * raw expression they are associated with.
+   *
+   * The output name of the original expression is assumed to be the desired final name
+   * of the rewritten expression, so if necessary an alias is added to ensure the output
+   * name is correct.
+   *
+   * @param ex expression to re-write
+   * @param aliasMap aliases from the input operator, mapped to their expressions
+   * @return rewritten expression with intermediate aliases removed
+   */
+  def resolveAliasesMaintainingSchema(ex: NamedExpression,
+                                      aliasMap: AttributeMap[Expression]): NamedExpression = {
+    val ret = replaceAlias(ex, aliasMap)
+    ret match {
+      case namedEx: NamedExpression =>
+        if (namedEx.name != ex.name) {
+          Alias(ret, ex.name)(ex.exprId, ex.qualifier, Some(ex.metadata))
+        } else {
+          namedEx
+        }
+      case _ => Alias(ret, ex.name)(ex.exprId, ex.qualifier, Some(ex.metadata))
+    }
+  }
+
+  def collapseIntoOneAggregate(aggExprs: Seq[NamedExpression],
+                               groupExprs: Seq[Expression],
+                               childAgg: Aggregate): Aggregate = {
+
+    val aliasMap = AttributeMap(childAgg.aggregateExpressions.collect {
+      case a: Alias => (a.toAttribute, a.child)
+    })
+    val aliasResolvedAggExprs = aggExprs.map(resolveAliasesMaintainingSchema(_, aliasMap))
+    val aliasResolvedGroupExprs = groupExprs.map(ex => replaceAlias(ex, aliasMap))
+    Aggregate(aliasResolvedGroupExprs, aliasResolvedAggExprs, childAgg.child)
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = {
+    plan transform {
+      // The query execution/optimization does not guarantee the expressions are evaluated in order.
+      // We only can combine them if and only if both are deterministic.
+      case agg@Aggregate(groupExprs: Seq[Expression],
```

Comment:
I think you can only collapse outer and inner Aggregate if the outer has grouping expression among inner's grouping expressions?

For example:

```
SELECT sum(agg), min(a), b from (
  SELECT sum(1) as agg, max(c) as maxC, a, b FROM testData2 GROUP BY a, b
) GROUP BY maxC
```
->
```
SELECT sum(1) as `sum(agg)`, max(c) as maxC, min(a) as `min(a)`, b from testData2 GROUP BY maxC
```
?

Reply:
I don't believe the collapse query is valid, it is referring to an aggregate expression alias in the group by, this one I think needs 2 phases of aggregates.

Reply:
Yea, it's not valid, but does current rule prevent it?

Reply:
It does avoid rewriting in these cases where the outer aggregate is grouping on the results of a grouping function of the inner aggregate. This test included negative case has a similar structure to your example.

```
  test("Don't collapse necessary nested aggregates, refer to inner agg result in outer group by") {
    val query = """
                  |SELECT sum(1), agg from (
                  |   select sum(1) as agg, a, b FROM testData2 GROUP BY a, b
                  |) group by agg
                """.stripMargin
```